kube.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	log "github.com/golang/glog"
+	"golang.org/x/net/context"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/rest"
+)
+
+// resetNodeCondition updates status of a Kubernetes node this flannel process
+// is running on to reset NetworkUnavailable condition
+//
+// On certain clouds, namely Google Compute Engine, nodes come up explicitly
+// marked with NetworkUnavailable condition which renders them unschedulable
+// and Kubernetes expects this condition to be removed either by RouteController
+// or networking plugins.
+//
+// We use neither of those: flannel does all route management (including
+// creating routes using cloud APIs using proper backend) and thus it has
+// to reset this condition itself.
+//
+// According to the current Kubernetes source code and multiple Github issues
+// this is an "expected" way to workaround this and what other people are
+// doing as well (https://github.com/kubernetes/kubernetes/issues/44254,
+// https://github.com/weaveworks/weave/issues/3249).
+func resetNodeCondition(ctx context.Context) {
+	ticker := time.NewTicker(retryInterval)
+	defer ticker.Stop()
+	// retry until we succeed because the node may not be registered yet
+	// (kubelet comes up after flannel)
+	for {
+		select {
+		case <-ticker.C:
+			client, err := getKubeClient()
+			if err != nil {
+				log.Warningf("Failed to create kube client: %v", err)
+				continue
+			}
+			patch, err := formatPatch()
+			if err != nil {
+				log.Warningf("Failed to format patch: %v", err)
+				continue
+			}
+			_, err = client.CoreV1().Nodes().PatchStatus(opts.kubeNode, patch)
```

Comment:
You'd only need to do this upon encountering `NodeNetworkUnavailable` (and not unconditionally) and hence only when the kubelet has actually started (since it's the one setting the status) - otherwise you run the risk of patching it before kubelet resets it.

Reply:
As I mentioned above, this status is only set during initial node registration. When kubelet starts for the first time, the node is not registered yet and hence this code will be spinning. Then kubelet will register the node which already includes NetworkUnavailable status, so it will be on the node right from the start.

Basically, what I'm trying to say is the race you're describing can't happen, so I don't think it makes sense to overcomplicate with trying to check the status first. If flannel restarts, it will update it again, yes, but that's fine - it's kind of like an "upsert".